org.testeditor.web.backend.persistence/src/test/java/org/testeditor/web/backend/persistence/DocumentProviderTest.xtend

```diff
@@ -18,12 +21,124 @@ class DocumentProviderTest extends AbstractGitTest {
+
 		// when
 		documentProvider.create('a.txt', 'test')
 
 		// then
-		val commits = git.log.call
-		commits.assertSize(2)
+		val numberOfCommitsAfter = git.log.call.size
+		numberOfCommitsAfter.assertEquals(numberOfCommitsBefore + 1)
+	}
+
+	@Test
+	def void createPushesChanges() {
+		// given
+		val numberOfCommitsBefore = remoteGit.log.call.size
+		val newFileName = "theNewFile.txt"
+
+		// when
+		documentProvider.create(newFileName, 'test')
+
+		// then
+		remoteGit.assertSingleCommit(numberOfCommitsBefore, ADD, newFileName)
 	}
 
+	@Test
+	def void createOrUpdateNonExistingFileAddsAndCommits() {
+		// given
+		val nonExistingFile = "aFileThatHasNotYetBeenCreated.txt"
+		val numberOfCommitsBefore = git.log.call.size
+
+		// when
+		documentProvider.createOrUpdate(nonExistingFile, "Contents of new file")
+
+		// then
+		git.assertSingleCommit(numberOfCommitsBefore, ADD, nonExistingFile)
+	}
+
+	@Test
+	def void createOrUpdatePreExistingFileCommitsModifications() {
+		// given
+		val preExistingFile = preExistingFileInRepo
+		val numberOfCommitsBefore = git.log.call.size
+
+		// when
+		documentProvider.createOrUpdate(preExistingFile, "New contents of pre-existing file")
+
+		// then
+		git.assertSingleCommit(numberOfCommitsBefore, MODIFY, preExistingFile)
+	}
+	
+	@Test
+	def void saveCommitsChanges() {
+		// given
+		val existingFileName = preExistingFileInRepo
+		val numberOfCommitsBefore = git.log.call.size
+
+		// when
+		documentProvider.save(existingFileName, "Contents of pre-existing file")
+
+		// then
+		git.assertSingleCommit(numberOfCommitsBefore, MODIFY, existingFileName)
+	}
+
+	@Test
+	def void savePushesChanges() {
+		// given
+		val existingFileName = preExistingFileInRepo
+		val numberOfCommitsBefore = remoteGit.log.call.size
+
+		// when
+		documentProvider.save(existingFileName, "Contents of pre-existing file")
+
+		// then
+		remoteGit.assertSingleCommit(numberOfCommitsBefore, MODIFY, existingFileName)
+	}
+
+	@Test
+	def void deleteCommitsChanges() {
+		// given
+		val existingFileName = preExistingFileInRepo
+		val numberOfCommitsBefore = git.log.call.size
+
+		// when
+		documentProvider.delete(existingFileName)
+
+		// then
+		git.assertSingleCommit(numberOfCommitsBefore, DELETE, existingFileName)
+	}
+
+	@Test
+	def void deletePushesChanges() {
+		// given
+		val existingFileName = preExistingFileInRepo
+		val numberOfCommitsBefore = remoteGit.log.call.size
+
+		// when
+		documentProvider.delete(existingFileName)
+
+		// then
+		remoteGit.assertSingleCommit(numberOfCommitsBefore, DELETE, existingFileName)
+	}
+
+	private def preExistingFileInRepo() {
+		val filename = "preExistingFile.txt"
+		localGitRoot.newFile(filename).createNewFile
+		git.pull.call
+		git.add.addFilepattern(filename).call
+		git.commit.setMessage("set test preconditions").call
+		git.push.call
+		return filename
+	}
+
+	private def assertSingleCommit(Git git, int numberOfCommitsBefore, ChangeType expectedChangeType, String path) {
```

Comment:
suggests that only the expectedChangeType is present, but checks whether at least one is present, right?

Reply:
hm, the "single" refers to a single commit, not a single diff entry. In my current test cases, I always have only a single diff entry, too, but you're right, it checks if there is at least one diff entry that satisfies the condition.

So… `assertSingleCommitContaining`, and then maybe reorder the arguments so it reads better? The `numberOfCommitsBefore` is still the odd one out, though.